src/Diagnostics/Samples/CSharp/CSharpAnalyzers/CSharpAnalyzers/StatefulAnalyzers/CodeBlockStartedAnalyzer.cs

```diff
@@ -0,0 +1,128 @@
+    /// <para>
+    /// The analyzer registers:
+    /// (a) A code block start action, which initializes per-code block mutable state. We mark all parameters as unused at start of analysis.
+    /// (b) A code block syntax node action, which identifes parameter references and marks the corresponding parameter as used.
+    /// (c) A code block end action, which reports diagnostics based on the final state, for all parameters which are unused.
+    /// </para>
+    /// </summary>
+    [DiagnosticAnalyzer(LanguageNames.CSharp)]
+    public class CodeBlockStartedAnalyzer : DiagnosticAnalyzer
+    {
+        #region Descriptor fields
+
+        internal static readonly LocalizableString Title = new LocalizableResourceString(nameof(Resources.CodeBlockStartedAnalyzerTitle), Resources.ResourceManager, typeof(Resources));
+        internal static readonly LocalizableString MessageFormat = new LocalizableResourceString(nameof(Resources.CodeBlockStartedAnalyzerMessageFormat), Resources.ResourceManager, typeof(Resources));
+        internal static readonly LocalizableString Description = new LocalizableResourceString(nameof(Resources.CodeBlockStartedAnalyzerDescription), Resources.ResourceManager, typeof(Resources));
+
+        internal static DiagnosticDescriptor Rule = new DiagnosticDescriptor(DiagnosticIds.CodeBlockStartedAnalyzerRuleId, Title, MessageFormat, DiagnosticCategories.Stateful, DiagnosticSeverity.Warning, isEnabledByDefault: true, description: Description);
+
+        #endregion
+
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(Rule);
+
+        public override void Initialize(AnalysisContext context)
+        {
+            context.RegisterCodeBlockStartAction<SyntaxKind>(startCodeBlockContext =>
+            {
+                // We only care about method bodies.
+                if (startCodeBlockContext.OwningSymbol.Kind != SymbolKind.Method)
+                {
+                    return;
+                }
+
+                // We only care about methods with parameters.
+                var method = (IMethodSymbol)startCodeBlockContext.OwningSymbol;
+                if (method.Parameters.IsEmpty)
+                {
+                    return;
+                }
+
+                // Initialize local mutable state in the start action.
+                var analyzer = new UnusedParametersAnalyzer(method);
+
+                // Register an intermediate non-end action that accesses and modifies the state.
+                startCodeBlockContext.RegisterSyntaxNodeAction(analyzer.AnalyzeSyntaxNode, SyntaxKind.IdentifierName);
+
+                // Register an end action to report diagnostics based on the final state.
+                startCodeBlockContext.RegisterCodeBlockEndAction(analyzer.CodeBlockEndAction);
+            });
+        }
+
+        private class UnusedParametersAnalyzer
```

Comment:
Sri, moving the interfaceType for CompilationStartedAnalyzer to a lambda local would mean the nested symbol action callback needs to be fully implemented within a nested lambda, rather then a separate method. This will make it inconsistent with all other sample analyzers + our default symbol analyzer in Analyzer + Fix project, where the core analysis callback is defined as a separate method. Is that fine?